includes/Core/Email_Reporting/Email_Log.php

```diff
@@ -0,0 +1,340 @@
+	 *
+	 * @since n.e.x.t
+	 */
+	protected function register_post_statuses() {
+		$statuses = array(
+			self::STATUS_SENT,
+			self::STATUS_FAILED,
+			self::STATUS_SCHEDULED,
+		);
+
+		foreach ( $statuses as $key => $status ) {
+			register_post_status(
+				$status,
+				array(
+					'public'                    => false,
+					'internal'                  => true,
+					'exclude_from_search'       => true,
+					'show_in_admin_all_list'    => false,
+					'show_in_admin_status_list' => false,
+				)
+			);
+		}
+	}
+
+	/**
+	 * Registers meta data for the email log post type.
+	 *
+	 * @since n.e.x.t
+	 */
+	protected function register_post_meta() {
+		$auth_callback = array( __CLASS__, 'meta_auth_callback' );
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_REPORT_FREQUENCY,
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_frequency' ),
+			)
+		);
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_BATCH_ID,
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_batch_id' ),
+			)
+		);
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_SEND_ATTEMPTS,
+			array(
+				'type'              => 'integer',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_attempts' ),
+			)
+		);
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_ERROR_DETAILS,
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_error_details' ),
+			)
+		);
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_REPORT_REFERENCE_DATES,
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_reference_dates' ),
+			)
+		);
+	}
+
+	/**
+	 * Sanitizes the report frequency meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_frequency( $value ) {
+		$allowed = array(
+			Reporting_Settings::FREQUENCY_WEEKLY,
+			Reporting_Settings::FREQUENCY_MONTHLY,
+			Reporting_Settings::FREQUENCY_QUARTERLY,
+		);
+		$value   = is_string( $value ) ? strtolower( $value ) : '';
+
+		return in_array( $value, $allowed, true ) ? $value : '';
+	}
+
+	/**
+	 * Sanitizes the batch ID meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_batch_id( $value ) {
+		$value = sanitize_text_field( (string) $value );
+
+		return substr( $value, 0, 191 );
```

Comment:
Let's define this length as a constant and document it, as this magic constant is a bit hard-to-read/make-sense-of on its own like this.